htmresearch/frameworks/pytorch/functions/k_winners.py

```diff
@@ -0,0 +1,207 @@
+    else:
+      boosted = x.detach()
+
+    # Take the boosted version of the input x, find the top k winners.
+    # Compute an output that contains the values of x corresponding to the top k
+    # boosted values
+    res = torch.zeros_like(x)
+    topk, indices = boosted.topk(k, sorted=False)
+    for i in range(x.shape[0]):
+      res[i, indices[i]] = x[i, indices[i]]
+
+    ctx.save_for_backward(indices)
+    return res
+
+
+  @staticmethod
+  def backward(ctx, grad_output):
+    """
+    In the backward pass we receive the context object and a Tensor containing
+    the gradient of the loss with respect to the output produced during the
+    forward pass. We can retrieve cached data from the context object, and must
+    compute and return the gradient of the loss with respect to the input to the
+    forward function.
+    """
+    indices, = ctx.saved_tensors
+    grad_x = torch.zeros_like(grad_output, requires_grad=True)
+
+    # Probably a better way to do it, but this is not terrible as it only loops
+    # over the batch size.
+    for i in range(grad_output.size(0)):
+      grad_x[i, indices[i]] = grad_output[i, indices[i]]
+
+    return grad_x, None, None, None
+
+
+
+class k_winners2d(torch.autograd.Function):
+  """
+  A K-winner take all autograd function for CNN 2D inputs (batch, Channel, H, W)
+  """
+
+
+  @staticmethod
+  def forward(ctx, x, dutyCycles, k, boostStrength):
+    """
+    :param ctx:
+      Place where we can store information we will need to compute the gradients
+      for the backward pass.
+
+    :param x:
+      Current activity of each unit.
+
+    :param dutyCycles:
+      The averaged duty cycle of each unit.
+
+    :param k:
+      The activity of the top k units will be allowed to remain, the rest are
+      set to zero.
+
+    :param boostStrength:
+      A boost strength of 0.0 has no effect on x.
+
+    :return:
+      A tensor representing the activity of x after k-winner take all.
+    """
+    batchSize = x.shape[0]
+    if boostStrength > 0.0:
+      targetDensity = float(k) / (x.shape[1] * x.shape[2] * x.shape[3])
+      boostFactors = torch.exp((targetDensity - dutyCycles) * boostStrength)
+      boosted = x.detach() * boostFactors
+    else:
+      boosted = x.detach()
+
+    # Take the boosted version of the input x, find the top k winners.
+    # Compute an output that only contains the values of x corresponding to the top k
+    # boosted values. The rest of the elements in the output should be 0.
+    boosted = boosted.reshape((batchSize, -1))
+    xr = x.reshape((batchSize, -1))
+    res = torch.zeros_like(boosted)
+    topk, indices = boosted.topk(k, dim=1, sorted=False)
+    res.scatter_(1, indices, xr.gather(1, indices))
+    res = res.reshape(x.shape)
+
+    ctx.save_for_backward(indices)
+    return res
+
+
+  @staticmethod
+  def backward(ctx, grad_output):
+    """
+    In the backward pass we receive the context object and a Tensor containing
```

Comment:
"In the backward pass, we set the gradient to 1 for the winning units, and 0 for the others."

Reply:
I just moved the k-winners file to the `functions` folder without looking at the comments. I also forgot to delete the original file in the parent folder. 
I will update the comments and delete the copy in the parent folder. 
Should I also delete the `linear_sdr` and `cnn_sdr` files? These files are now obsolete.

Reply:
Yes. Delete all the (old) things!